### PR TITLE
#182461139 - USERS SHOULD BE ABLE  TO RESET PASSWORD VIA EMAIL

### PR DESCRIPTION
**Pull request by ChrisOnsando**
#### What does this Pull Request do?
- [ ] Users should be able to reset their password via email
#### Description of Task to be completed?
- [ ] Users should be able to reset password via email
#### How should this be manually tested?
- [ ] N/A
#### Any background context you want to provide?
- [ ] Users should be able to reset their password via email
#### What are the relevant Pivotal Tracker Story?
- [ ] https://www.pivotaltracker.com/story/show/182461139
#### Screenshots (if appropriate)
- [ ] N/A
#### Questions:
- [ ] None

### CODE REVIEW

**Comment by dalienst** [comment]
The PR is ready for merging

File: Pipfile [hunk 1]
```diff
@@ -12,6 +12,10 @@ whitenoise = "6.2.0"
 gunicorn = "20.1.0"
 djangorestframework = "3.13.1"
 drf-yasg = "1.20.0"
+typing-extensions = "4.2.0"
+djangorestframework-simplejwt = "*"
+mypy = "*"
```
Comment (mainamuriuki274) [inline Pipfile:17]: should be in dev packages and have version

File: Pipfile [hunk 2]
```diff
@@ -12,6 +12,10 @@ whitenoise = "6.2.0"
 gunicorn = "20.1.0"
 djangorestframework = "3.13.1"
 drf-yasg = "1.20.0"
+typing-extensions = "4.2.0"
```
Comment (mainamuriuki274) [inline Pipfile:15]: No need for this python has an inbuilt typing library

File: Pipfile [hunk 3]
```diff
@@ -12,6 +12,10 @@ whitenoise = "6.2.0"
...
+djangorestframework-simplejwt = "*"
```
Comment (mainamuriuki274) [inline Pipfile:16]: version number?

File: app/users/serializers.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    )
+
+    class Meta:
+
+        fields = ["password"]
+
+    def validate(self, data: Any) -> Any:
+
+        password = data.get("password")
+        token = self.context.get("kwargs").get("token")  # type: ignore
```
Comment (mainamuriuki274) [inline app/users/serializers.py:73]: don't ignore generically ignore only iggnore specific error codes

File: app/users/serializers.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+        if token is None or encoded_pk is None:
+            raise serializers.ValidationError("Missing data.")
+
+        pk = urlsafe_base64_decode(encoded_pk).decode()
+        user = User.objects.get(pk=pk)
+        if not PasswordResetTokenGenerator().check_token(user, token):
+            raise serializers.ValidationError("The reset token is invalid")
+
+        user.set_password(password)
```
Comment (mainamuriuki274) [inline app/users/serializers.py:84]: Override the  drf save method and put line 84, 85 there

File: app/users/tests/test_views.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        self.factory = RequestFactory()
+        self.user = User.objects.create_user(
+            username="chrisnemwel",
+            email="chrisnemwel@gmail.com",
+            password="secretpassword",
+        )
+
+    def test_password_reset_email(self) -> None:
+        url = reverse("request-password-reset")
+        factory = RequestFactory()
```
Comment (mainamuriuki274) [inline app/users/tests/test_views.py:24]: Haven't you already created a factory for each request in the setUp method

File: Pipfile [hunk 7]
```diff
@@ -12,6 +12,10 @@ whitenoise = "6.2.0"
...
+mypy = "*"
+django-rest-passwordreset = "*"
```
Comment (mainamuriuki274) [inline Pipfile:18]: This package is not actively maintained, avoid using third party packages where possible, it might pose security risks or incompatibility issues in the future, you can ask  Dalienst how he override the Django PasswordTokenGenerator class

File: app/users/serializers.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    def validate(self, attrs: Any) -> Any:
+        try:
+            request = self.context.get("request")
+            email = attrs.get("email")
+            user = Util.generate_reset_token(email)
+
+            if user:
+                email_data = Util.create_reset_email(request, *user)
+                Util.send_email("password_reset.html", email_data)
+        except KeyError:
```
Comment (mainamuriuki274) [inline app/users/serializers.py:54]: No need for this if you have the email required=True

**Review by mainamuriuki274 — CHANGES_REQUESTED** [review]

File: Pipfile (hunk 2 above)
Comment (ChrisOnsando) [reply Pipfile:15]: Okay

File: Pipfile (hunk 3 above)
Comment (ChrisOnsando) [reply Pipfile:16]: I have updated with the version number

File: app/users/serializers.py (hunk 4 above)
Comment (ChrisOnsando) [reply app/users/serializers.py:73]: I get error in mypy when I dont ignore this

File: app/users/serializers.py (hunk 5 above)
Comment (ChrisOnsando) [reply app/users/serializers.py:84]: I have experienced a challenge doing this

File: app/users/tests/test_views.py (hunk 6 above)
Comment (ChrisOnsando) [reply app/users/tests/test_views.py:24]: What if I remove the create factory in the setUp method , Is it proper to create a factory for each request

File: app/users/serializers.py [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+            raise ParseError("email must be provided")
+        return super().validate(attrs)
+
+
+class ResetPasswordSerializer(serializers.Serializer):
+
+    password = serializers.CharField(
+        write_only=True,
+        min_length=1,
```
[the remark this answers is not in the document]
Comment (ChrisOnsando) [reply app/users/serializers.py:63]: I have increased this

File: app/users/views.py [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+
+class PasswordReset(generics.GenericAPIView):
+    """
+    Request for Password Reset Link.
+    """
+
+    serializer_class = PasswordResetSerializer
+
+    def post(self, request: Any) -> Any:
```
[the remark this answers is not in the document]
Comment (ChrisOnsando) [reply app/users/views.py:37]: Change made

File: app/users/urls.py [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+
+from app.users import views
+from app.users.views import UserRegister
+
+urlpatterns = [
+    path("token/", TokenObtainPairView.as_view(), name="token_obtain_pair"),
+    path("token/refresh/", TokenRefreshView.as_view(), name="token_refresh"),
+    path("register/", UserRegister.as_view(), name="register"),
+    path(
+        "",
```
[the remark this answers is not in the document]
Comment (ChrisOnsando) [reply app/users/urls.py]: Path added

**Comment by ChrisOnsando** [comment]
All the changes requested made

File: Pipfile [hunk 12]
```diff
@@ -12,11 +12,13 @@ whitenoise = "6.2.0"
 gunicorn = "20.1.0"
 djangorestframework = "3.13.1"
 drf-yasg = "1.20.0"
+faker = "13.14.0"
```
Comment (mainamuriuki274) [inline Pipfile:15]: this is should be a dev dependancy

File: app/users/models.py [hunk 13]
```diff
@@ -84,3 +85,12 @@ class User(AbstractBaseUser, PermissionsMixin, TimeStampedModel):
     objects = UserManager()
     REQUIRED_FIELDS = ["username", "password"]
     USERNAME_FIELD = "email"
+
+
+class Profile(models.Model):
```
Comment (mainamuriuki274) [inline app/users/models.py:90]: Remove the profile model because this is another card

File: app/users/serializers.py [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+from app.users.utils import (
+    create_reset_email,
+    generate_reset_token,
+    send_email,
+)
+
+User = get_user_model()
+
+
+class UserSerializer(serializers.ModelSerializer):
```
Comment (mainamuriuki274) [inline app/users/serializers.py:18]: Why do you have this yet it belongs to another story?
it will cause conflicts later

File: app/users/serializers.py (hunk 5 above)
Comment (mainamuriuki274) [reply app/users/serializers.py:88]: Why did you not ask me? or you team members? this is team work

File: app/users/tests/test_views.py [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+from django.test import RequestFactory, TestCase
+from django.urls import reverse
+from django.utils.encoding import force_bytes
+from django.utils.http import urlsafe_base64_encode
+from faker import Faker
+from rest_framework import status
+
+from app.users.views import PasswordReset, ResetPasswordAPI
+
+password_reset = PasswordReset.as_view()
```
Comment (mainamuriuki274) [inline app/users/tests/test_views.py:12]: remove views and factory from this file

File: app/users/tests/test_views.py [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+        )
+
+    def test_password_reset_request(self) -> None:
+        url = reverse("request-password-reset")
+        data = {"email": fake.email()}
+        response = self.client.post(url, data, format="json")
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+    def test_password_reset_no_email(self) -> None:
+        factory = RequestFactory()
```
Comment (mainamuriuki274) [inline app/users/tests/test_views.py:31]: remove factory from this file

File: app/users/tests/test_views.py [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+        url = reverse("request-password-reset")
+        request = factory.post(url, format="json")
+        response = password_reset(request)
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
```
Comment (mainamuriuki274) [inline app/users/tests/test_views.py:35]: assert also for the message

File: app/users/tests/test_views.py [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+        )
+
+        response = self.client.post(
+            reset_url, data={"password": "mypassword"}, format="json"
+        )
+
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+        self.assertIn(
+            "The reset token is invalid", str(response.data["detail"])  # type: ignore[attr-defined]
+        )
```
Comment (mainamuriuki274) [inline app/users/tests/test_views.py:72]: Add other edge cases eg if invalid uidb64 is supplied

File: app/users/urls.py [hunk 19]
```diff
@@ -0,0 +1,19 @@
...
+    TokenObtainPairView,
+    TokenRefreshView,
+)
+
+from app.users.views import PasswordReset, ResetPasswordAPI, UserRegister
+
+urlpatterns = [
+    path("token/", TokenObtainPairView.as_view(), name="token_obtain_pair"),
+    path("token/refresh/", TokenRefreshView.as_view(), name="token_refresh"),
+    path("api/register/", UserRegister.as_view(), name="register"),
```
Comment (mainamuriuki274) [inline app/users/urls.py:12]: you are repeating api, isn't  /api this in the main urls.py

File: app/users/urls.py [hunk 20]
```diff
@@ -0,0 +1,19 @@
...
+    path("", PasswordReset.as_view(), name="request-password-reset"),
```
Comment (mainamuriuki274) [inline app/users/urls.py:13]: add pathname to this path

File: app/users/views.py [hunk 21]
```diff
@@ -0,0 +1,68 @@
...
+        response["refresh"] = str(refresh)
+        response["access"] = str(refresh.access_token)
+
+        return Response(response, status=status.HTTP_201_CREATED)
+
+
+class PasswordReset(generics.GenericAPIView):
+    """
+    Request for Password Reset Link.
+
```
Comment (mainamuriuki274) [inline app/users/views.py:33]: remove this empty line

File: app/users/views.py [hunk 22]
```diff
@@ -0,0 +1,68 @@
...
+
+class ResetPasswordAPI(generics.GenericAPIView):
+
+    serializer_class = ResetPasswordSerializer
+
+    def post(self, request: Any, *args: Any, **kwargs: Any) -> Any:
+        """
+        Verify token & encoded_pk and then reset the password.
+
```
Comment (mainamuriuki274) [inline app/users/views.py:57]: remove this empty line

File: app/users/views.py [hunk 23]
```diff
@@ -0,0 +1,68 @@
...
+        """
+        serializer = self.serializer_class(
```
Comment (mainamuriuki274) [inline app/users/views.py:59]: use self.get_serializer instead of self.serializer_class

**Review by mainamuriuki274 — CHANGES_REQUESTED** [review]

File: app/users/views.py (hunk 22 above)
Comment (ChrisOnsando) [reply app/users/views.py:57]: Change made

File: app/users/views.py (hunk 23 above)
Comment (ChrisOnsando) [reply app/users/views.py:59]: What is the difference between the two ?

File: app/users/serializers.py (hunk 14 above)
Comment (ChrisOnsando) [reply app/users/serializers.py:18]: I have this because we need to authenticate the user, for it to be possible to reset their password

File: app/users/tests/test_views.py (hunk 15 above)
Comment (ChrisOnsando) [reply app/users/tests/test_views.py:12]: I have made the changes

File: app/users/tests/test_views.py (hunk 16 above)
Comment (ChrisOnsando) [reply app/users/tests/test_views.py:31]: I have removed the factory

File: app/users/views.py (hunk 21 above)
Comment (ChrisOnsando) [reply app/users/views.py:33]: Okay

File: app/users/urls.py (hunk 20 above)
Comment (ChrisOnsando) [reply app/users/urls.py:13]: Path added

File: Pipfile [hunk 24]
```diff
@@ -17,13 +17,15 @@ drf-yasg = "1.20.0"
...
 black = "22.3.0"
+djangorestframework-simplejwt = "5.2.0"
 isort = "5.10.1"
 pytest = "7.1.2"
 django-stubs = "1.11.0"
 mypy = "0.950"
 pytest-django = "4.5.2"
 pytest-cov = "3.0.0"
 djangorestframework-stubs = "1.7.0"
+faker = "13.14.0"
```
Comment (mainamuriuki274) [inline Pipfile:28]: this is a dev dependancy remove it

**Review by mainamuriuki274 — CHANGES_REQUESTED** [review]

File: app/users/urls.py (hunk 19 above)
Comment (ChrisOnsando) [reply app/users/urls.py]: Change made

File: Pipfile (hunk 7 above)
Comment (ChrisOnsando) [reply Pipfile]: Okay

File: app/users/tests/test_views.py [hunk 25]
```diff
@@ -0,0 +1,45 @@
...
+    def test_password_reset_email(self) -> None:
+        url = reverse("request-password-reset")
+        factory = RequestFactory()
+        request = factory.post(url, {"email": self.user.email}, format="json")
+        response = PasswordReset.as_view()(request)
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+    def test_request_password_reset(self) -> None:
+        url = reverse("request-password-reset")
+        data = {"email": "chrisnemwel@gmail.com", "password": "secretpassword"}
```
[the remark this answers is not in the document]
Comment (ChrisOnsando) [reply app/users/tests/test_views.py]: Implemented the faker library

File: app/users/tests/test_views.py [hunk 26]
```diff
@@ -0,0 +1,45 @@
...
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+    def test_password_reset_token(self) -> None:
+        factory = RequestFactory()
+        encoded_pk = urlsafe_base64_encode(force_bytes(self.user.pk))
+        url = reverse(
+            "reset-password", kwargs={"encoded_pk": encoded_pk, "token": None}
+        )
+        request = factory.patch(url)
+        response = ResetPasswordAPI.as_view()(
```
[the remark this answers is not in the document]
Comment (ChrisOnsando) [reply app/users/tests/test_views.py]: Okay

File: speaksfer/settings/base.py [hunk 27]
```diff
@@ -41,18 +39,18 @@
...
-    # third-party imports
-    "drf_yasg",
+    "app.users",
     "rest_framework",
-    # app imports
-    "app.authentication",
+    "rest_framework_simplejwt",
+    "rest_framework_simplejwt.token_blacklist",
+    "rest_framework.authtoken",
+    "django_rest_passwordreset",
```
[the remark this answers is not in the document]
Comment (ChrisOnsando) [reply speaksfer/settings/base.py]: Okay

**Comment by dalienst** [comment]
PR is ready for merging

File: app/user/serializers.py [hunk 28]
```diff
@@ -1,116 +1,72 @@
...
-        return user
+        return super().validate(attrs)
 
 
-class VerifyEmailSerializer(serializers.Serializer):
-    token = serializers.CharField()
-    uidb64 = serializers.CharField()
+class VerifyPasswordResetSerializer(serializers.Serializer):
+
+    password = serializers.CharField(
```
Comment (mainamuriuki274) [inline app/user/serializers.py:35]: Add password validators like during registering here so that the passwoed is strong

File: app/user/serializers.py [hunk 29]
```diff
@@ -1,116 +1,72 @@
...
-        except (TypeError, ValueError, OverflowError, User.DoesNotExist):
+        password = data.get("password")
+        url_kwargs = self.context.get("kwargs")
+        token = url_kwargs.get("token")  # type: ignore[union-attr]
+        encoded_pk = url_kwargs.get("encoded_pk")  # type: ignore[union-attr]
+
+        if token is None or encoded_pk is None:
             raise serializers.ValidationError(
-                "Invalid user id", code="invalid_code"
+                {"detail": "This field is required"}
```
Comment (mainamuriuki274) [inline app/user/serializers.py:53]: this error is not specific it does not say which one is required, and I do not think it is necessary (the if statement)

File: app/user/utils.py [hunk 30]
```diff
@@ -0,0 +1,53 @@
...
+def generate_reset_token(email: str) -> Any:
+    if User.objects.filter(email=email).exists():
+        user = User.objects.get(email=email)
+        encoded_pk = urlsafe_base64_encode(smart_bytes(user.id))
+        token = PasswordResetTokenGenerator().make_token(user)
+        return (user, encoded_pk, token)
+    return None
+
+
+def create_reset_email(
```
Comment (mainamuriuki274) [inline app/user/utils.py:38]: This function is not reusable, can only be used for reset, it should also be  able to be used for verify email functionality, also strrive to have your own implementations, this is just Brion's code copy paste

**Review by mainamuriuki274 — CHANGES_REQUESTED** [review]

File: Pipfile [hunk 31]
```diff
@@ -13,12 +13,12 @@ gunicorn = "20.1.0"
 djangorestframework = "3.13.1"
 drf-yasg = "1.20.0"
 djangorestframework-simplejwt = "5.2.0"
-django-hashid-field = "3.3.5"
 
 [dev-packages]
 pre-commit = "2.19.0"
 flake8 = "4.0.1"
 black = "22.3.0"
+djangorestframework-simplejwt = "5.2.0"
```
[the remark this answers is not in the document]
Comment (ChrisOnsando) [reply Pipfile:21]: Okay

[the remark this answers is not in the document]
Comment (ChrisOnsando) [reply Pipfile:21]: Having to change that

File: app/user/serializers.py (hunk 28 above)
Comment (ChrisOnsando) [reply app/user/serializers.py:35]: Okay

File: app/user/utils.py (hunk 30 above)
Comment (ChrisOnsando) [reply app/user/utils.py:38]: Will change that

File: app/user/serializers.py [hunk 32]
```diff
@@ -1,43 +1,47 @@
...
-        validators=[UniqueValidator(queryset=User.objects.all())],
-    )
+
+class VerifyPasswordResetSerializer(serializers.Serializer):
 
     password = serializers.CharField(
-        max_length=128,
-        min_length=8,
         write_only=True,
+        min_length=6,
```
Comment (mainamuriuki274) [inline app/user/serializers.py:44]: this length is too short for a password

**Review by mainamuriuki274 — CHANGES_REQUESTED** [review]

File: Pipfile [hunk 33]
```diff
@@ -13,7 +13,6 @@ gunicorn = "20.1.0"
 djangorestframework = "3.13.1"
 drf-yasg = "1.20.0"
 djangorestframework-simplejwt = "5.2.0"
-django-hashid-field = "3.3.5"
```
[the remark this answers is not in the document]
Comment (ChrisOnsando) [reply Pipfile:16]: Change made

File: app/user/templates/email_verification.html [hunk 34]
```diff
@@ -1,5 +1,11 @@
 {% autoescape off %}
-Hi {{ user.username }},
-Please click on the link to confirm your registration,
-http://{{ domain }}{% url 'email-verify' uidb64=uid token=token %}
+
+Hello {{ user.username }},
+
+Verify your email  by clicking the link below:
+{% block reset_link %}
+https://speaksfer/hi
```
[the remark this answers is not in the document]
Comment (ChrisOnsando) [reply app/user/templates/email_verification.html:7]: Change made

File: app/user/tests/test_views.py [hunk 35]
```diff
@@ -9,93 +10,98 @@
...
         }
+        cls.user = User.objects.create_user(**cls.testuser)
+        return super().setUpClass()
+
+    def test_password_reset_request(self) -> None:
+        url = reverse("password-reset")
+        data = {"email": self.testuser["email"]}
+        response = self.client.post(url, data, format="json")
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(len(mail.outbox), 1)
```
[the remark this answers is not in the document]
Comment (ChrisOnsando) [reply app/user/tests/test_views.py:36]: Done

**Comment by dalienst** [comment]
PR ready for merging
